src/main/java/uk/ac/ebi/ampt2d/accession/common/accessioning/BasicAccessioningService.java

```diff
@@ -0,0 +1,136 @@
+        this.dbService = dbService;
+        this.digestFunction = digestFunction;
+        this.hashingFunction = hashingFunction;
+    }
+
+    /**
+     * Get accessions for a list of messages. It looks for the object's accessions in a repository, and it they don't
+     * exist, generate new ones, storing them in the repository
+     *
+     * @param messages
+     * @return
+     */
+    @Override
+    public Map<ACCESSION, MODEL> getOrCreateAccessions(List<? extends MODEL> messages) {
+        Map<HASH, MODEL> hashToMessages = mapHashOfMessages(messages);
+        Map<HASH, ACCESSION> existingAccessions = dbService.getExistingAccessions(hashToMessages.keySet());
+        Map<HASH, MODEL> newMessages = filterNotExists(hashToMessages, existingAccessions);
+
+        Map<ACCESSION, MODEL> accessions = joinExistingAccessionsWithMessages(existingAccessions, hashToMessages);
+        if (!newMessages.isEmpty()) {
+            accessions.putAll(generateAccessions(newMessages));
+        }
+        return accessions;
+    }
+
+    /**
+     * Digest messages, hash them and map them. If Two messages have the same hash keep the first one.
+     *
+     * @param messages
+     * @return
+     */
+    private Map<HASH, MODEL> mapHashOfMessages(List<? extends MODEL> messages) {
+        return messages.stream().collect(Collectors.toMap(digestFunction.andThen(hashingFunction), e -> e, (r, o) -> r));
+    }
+
+    private Map<HASH, MODEL> filterNotExists(Map<HASH, MODEL> hashToMessages,
+                                             Map<HASH, ACCESSION> existingAccessions) {
+        return hashToMessages.entrySet().stream()
+                .filter(entry -> !existingAccessions.containsKey(entry.getKey()))
+                .collect(Collectors.toMap(entry -> entry.getKey(), entry -> entry.getValue()));
+    }
+
+    private Map<ACCESSION, MODEL> joinExistingAccessionsWithMessages(Map<HASH, ACCESSION> existingAccessions,
+                                                                     Map<HASH, MODEL> hashToMessages) {
+        return existingAccessions.values().stream().collect(Collectors.toMap(e -> e, e -> hashToMessages.get(e)));
+    }
+
+    private Map<ACCESSION, MODEL> generateAccessions(Map<HASH, MODEL> accessions) {
+        SaveResponse<ACCESSION, MODEL> response = dbService.save(accessionGenerator.generateAccessions(accessions));
+        accessionGenerator.postSave(response);
+        return response.getAllAccessionToMessage();
+    }
+
+    @Override
+    public Map<ACCESSION, MODEL> getAccessions(List<? extends MODEL> accessionObjects) {
+        return dbService.findAllAccessionByMessageHash(getHashes(accessionObjects));
+    }
+
+    private List<HASH> getHashes(List<? extends MODEL> accessionObjects) {
+        return accessionObjects.stream().map(digestFunction.andThen(hashingFunction)).collect(Collectors.toList());
+    }
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        // Do nothing
+    }
```

Comment:
am not sure whats the use of empty afterPropertiesSet InitializingBean

Reply:
The major need is the child class of this that needs to initialize code, didn't see harm in enforcing that all child classes of BasicAccessioningService implement the InitializingBean interface to ensure that everybody understands that this cass is suppossed to be used injected and that initialization after injection was something possible.

Still, the implement can be moved to the child only, any stronger opinion?

Reply:
I have commented in another part of the code, that BasicAccessioningService (ant the afterPropertiesSet method) could be made abstract, and then you force the children to implement it (even if they leave it empty, but at least somebody extending this class is aware of that)